accepted/wasmtime-baseline-compilation.md

```diff
@@ -0,0 +1,260 @@
+introduction of a baseline compiler.
+
+**Approximate** [measurements](https://github.com/Shopify/wasm-bench) taken on
+top of a subset of the [Sightglass benchmark
+suite](https://github.com/bytecodealliance/sightglass/tree/main/benchmarks)
+– using different optimizing and baseline compilers (Cranelift from Wasmtime,
+Liftoff from V8 and RabaldrMonkey from SpiderMonkey) – show that a baseline
+compiler on average yields 15x to 20x faster compilation while producing code
+that is on average 1.1x to 1.5x slower than the one produced by an optimizing
+compiler. These measurements align on average with other measurements observed
+when comparing interpretation and compilation for WebAssembly[^1].
+
+
+[^1]: Ben L. Titzer. [A fast in-place interpreter for
+  WebAssembly](https://arxiv.org/pdf/2205.01183.pdf)
+
+## Proposal: Winch, a baseline compiler for Wasmtime
+
+Winch: WebAssembly Intentionally-Non-Optimizing Compiler and Host
+
+
+### Design Principles
+
+* Single pass over Wasm bytecode
+* Function as the unit of compilation
+* Machine code generation directly from Wasm bytecode – no intermediate
+  representation
+* Avoid reinventing machine-code emission – use Cranelift's instruction emitter
+  code to create an assembler library ("MacroAssembler")
+* Prioritize compilation performance over runtime performance
+
+### High-level overview
+
+```mermaid
+  graph TD;
+  A(wasmparser)-->B(cranelift-wasm);
+  A-->C(winch);
+  C-->D(MacroAssembler);
+  D-->X(cranelift-asm);
+  X-->E(MachInst);
+  X-->F(MachBuffer);
+  B-->G(cranelift);
+  G-->X;
+```
+
+### MacroAssembler and Borrowing from Cranelift
+
+We plan to factor out the lower layers of Cranelift that produce and operate on
+machine code in order to reuse them as a generic assembler library
+(“MacroAssembler”).
+
+The two key abstractions that will be useful to reuse are the `MachInst`
+(“machine instruction”) trait and its implementations for each architecture; and
+the `MachBuffer`, which is a machine-code emission buffer with some knowledge of
+branches and ability to do peephole optimizations on them. The former lets us
+reuse all the logic to encode instructions for an ISA; the latter lets us emit
+code with “labels” and references to labels, and have the fixups done for us.
+
+The `MachInst` trait and its implementations, and the `MachBuffer`, can be
+mostly factored out into a separate crate `cranelift_asm`. This will require
+some care with respect to layering: in particular, definitions of
+machine-instruction types are currently done in the ISLE backends for each ISA
+within Cranelift. We can continue to use ISLE for these, but they will need to
+be moved to the separate crate.
+
+
+As a result of this initial layering, one will be able to build a `MachInst` as
+a Rust data structure and emit it manually, for example:
+
+```rust
+let add = cranelift_asm::x64::AluRmiR { op: AluRmiR::Add, … }; let mut
+buf = MachBuffer::new(); add.emit(&mut buf, …);
```

Comment:
Tiny linebreak suggestion: 

```suggestion
let add = cranelift_asm::x64::AluRmiR { op: AluRmiR::Add, … };
let mut buf = MachBuffer::new(); 
add.emit(&mut buf, …);
```

Reply:
I've bundled all the typo corrections in https://github.com/bytecodealliance/rfcs/pull/28/commits/d8d934ec8a16c6d0ae823c9d5ba499f472bfab99, thanks!